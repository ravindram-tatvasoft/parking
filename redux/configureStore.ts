import { combineReducers, createStore } from 'redux'
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage' // defaults to localStorage for web
import parkingSlotSlice from './reducer/parking-slot-slice'
import AsyncStorage from '@react-native-async-storage/async-storage';


const persistConfig = {
    key: 'root',
    storage: AsyncStorage,
}

const rootReducer = combineReducers({
    parkingSlot: parkingSlotSlice,
})

const persistedReducer = persistReducer(persistConfig, rootReducer)

let store = createStore(persistedReducer)
let persistor = persistStore(store)

export {
    store, persistor
}
