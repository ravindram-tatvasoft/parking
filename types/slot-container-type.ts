import { SlotListType } from "./slot-list-tyle";

export interface SlotContainerType {
    item: SlotListType,
    index: number
    onRemoveCar: (index:number, item: SlotListType) => void
}