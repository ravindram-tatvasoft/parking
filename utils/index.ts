export * from './strings'
export * from './colors'
export * from './common-style'
export * from './utils'