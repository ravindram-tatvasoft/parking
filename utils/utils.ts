import moment from 'moment'

const msToTime = (milli: number) => {
  return moment(milli).format("HH:mm:ss")
}

const getDifferenceDateString = (startTime: string, endTime: string): string => {
  var startDate = moment(startTime);
  var endDate = moment(endTime);

  var totalSec = endDate.diff(startDate, 'seconds');
  const time = moment().startOf('day')
        .seconds(totalSec)
        .format('HH@mm').split("@");
    if(parseInt(time[0]) > 0){
      return `${time[0]} hours ${time[1]} minutes`
    } else{
      return `${time[1]} minutes`
    }
}

const validateCarNo = (carNo: string): boolean => {
  var pattern = /[A-Z][A-Z] [0-9][0-9] [A-Z][A-Z] [0-9][0-9][0-9][0-9]$/i;
  var re = new RegExp(pattern);
  return re.test(carNo)
}

const getDifferenceDate = (startTime: string, endTime: string): any => {
  var startDate = moment(startTime);
  var endDate = moment(endTime);

  const h = endDate.diff(startDate, 'hours')
  const m = endDate.diff(startDate, 'minutes')
  return { hour: h, minute: m }
}

export {
  msToTime,
  getDifferenceDate,
  getDifferenceDateString,
  validateCarNo
}