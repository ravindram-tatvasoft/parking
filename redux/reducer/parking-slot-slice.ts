import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { SlotListType } from '../../types/slot-list-tyle'

export interface CounterState {
  totalParking: number
  slotList: SlotListType[]
}

const initialState: CounterState = {
    totalParking: 0,
    slotList: []
}

export const parkingSlotSlice = createSlice({
  name: 'parkingSlot',
  initialState,
  reducers: {
    setTotalParking: (state, action) => {
      state.totalParking = action.payload
    },
    setParkingSlotList: (state, action) => {
      state.slotList = action.payload
    },
  },
})

// Action creators are generated for each case reducer function
export const { setTotalParking, setParkingSlotList  } = parkingSlotSlice.actions

export default parkingSlotSlice.reducer