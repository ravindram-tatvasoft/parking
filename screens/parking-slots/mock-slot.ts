import moment from "moment";
import { SlotListType } from "../../types/slot-list-tyle";

export const mockDataSlots: SlotListType[] = [
    {
        id: 'id-1',
        exitTime: moment().format(),
        isEmpty: false,
        slotNumber: 1,
        time: 1,
        carNumber: "GJ 01 XX XXXX"
    },
    {
        id: 'id-2',
        exitTime: moment().format(),
        isEmpty: true,
        slotNumber: 2,
        time: 2,
        carNumber: "GJ 02 XX XXXX"
    }
]
