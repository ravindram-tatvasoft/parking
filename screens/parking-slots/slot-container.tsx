import { Dimensions, Image, StyleSheet, Text, TouchableOpacity, View } from "react-native"
import { images } from "../../assets/images/images"
import { SlotContainerType } from "../../types/slot-container-type"
import { SlotListType } from "../../types/slot-list-tyle"
import { colors } from "../../utils"

const width = Dimensions.get('screen').width
const viewSize = (width / 2) - 20


export const SlotContainer = (props: SlotContainerType) => {
    const { isEmpty, slotNumber } = props.item
    const parkingStatus = (item: SlotListType): string => {
        return `Slot ${slotNumber}\n ${item.isEmpty ? "Available" : "Not Available"}`
    }

    return (<TouchableOpacity disabled={isEmpty}
        style={styles.root}
        testID={`post-row-${props.index}`}
        onLongPress={() => props.onRemoveCar(props.index, props.item)}>
        <Text style={[styles.txtSlot, { color: isEmpty ? colors.green : colors.red }]}>{parkingStatus(props.item)}</Text>
        {!isEmpty && <View style={styles.carImgAndNumber}>
            <Text>{props.item.carNumber}</Text>
            <Image source={images.car} style={styles.carIcon} />
        </View>}
    </TouchableOpacity>)
}

const styles = StyleSheet.create({
    carImgAndNumber: {
        alignItems: "center"
    },
    root: {
        width: viewSize,
        height: viewSize,
        backgroundColor: colors.greyLight,
        margin: 10,
        alignItems: "center",
        justifyContent: 'center'
    },
    carIcon: {
        width: 50,
        height: 50,
        marginVertical: 10,
    },
    txtSlot: {
        textAlign: "center"
    }
})