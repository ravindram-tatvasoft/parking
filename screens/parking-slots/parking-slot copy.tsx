import { useEffect, useState } from 'react'
import { Alert, FlatList, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { commonStyles, string } from '../../utils';
import { SlotContainer } from './slot-container';
import 'react-native-get-random-values';
import { v4 as uuidv4 } from 'uuid';
import { SlotListType } from '../../types/slot-list-tyle';
import { useDispatch, useSelector } from 'react-redux';
import { styles } from '../createlot/create-lot.style';
import { setParkingSlotList } from '../../redux/reducer/parking-slot-slice';
import { useNavigation } from '@react-navigation/core';
import moment from 'moment'


export default function ParkingSlot({ navigation }) {

  const parkingSlot = useSelector(state => state.parkingSlot);
  const dispatch = useDispatch()

  const removeCarAlert = (index: number, item: SlotListType) => {
    Alert.alert('', 'Are you sure you want to remove car from parking?', [
      { text: 'Cancel', onPress: () => null },
      { text: 'OK', onPress: () => removeCar(index, item) },
    ], { cancelable: false })
  }

  const removeCar = (index: number, item: SlotListType) => {
    let updateCar: SlotListType = { ...parkingSlot.slotList[index] };
    updateCar.exitTime = moment().format()
    const tempList = JSON.parse(JSON.stringify(parkingSlot.slotList));
    tempList.splice(index, 1, updateCar)
    dispatch(setParkingSlotList(tempList))
    navigation.navigate("PaymentScreen", { slotDetail: updateCar })
  }

  const renderItem = ({ item, index }) => (
    <SlotContainer item={item} index={index} onRemoveCar={removeCarAlert} />
  );

  return (
    <View style={commonStyles.container}>
      {parkingSlot.slotList.length > 0 ? <View>
        <Text>{string.noteRemoveCar}</Text>
        <FlatList
          data={parkingSlot.slotList}
          renderItem={renderItem}
          numColumns={2}
          keyExtractor={item => item.id}
        />
      </View> : <View style={commonStyles.viewCenter}>
        <Text>{string.emptySlot}</Text>
      </View>}
    </View>
  );
}

