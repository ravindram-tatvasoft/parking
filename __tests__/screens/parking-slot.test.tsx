import { cleanup, render, fireEvent, waitFor, waitForElement } from "@testing-library/react-native";
import * as React from "react";
import { Alert } from 'react-native'
import { Provider } from "react-redux"
import renderer from 'react-test-renderer'
import { store } from "../../redux/configureStore"
import { setParkingSlotList } from "../../redux/reducer/parking-slot-slice";
import { mockDataSlots } from "../../screens/parking-slots/mock-slot";
import ParkingSlot from "../../screens/parking-slots/parking-slot"
import { SlotContainer } from "../../screens/parking-slots/slot-container";
import { SlotContainerType } from "../../types/slot-container-type";
import { string } from "../../utils";

jest.mock('@react-navigation/native', () => ({
    ...jest.requireActual('@react-navigation/native'),
    useNavigation: () => ({ goBack: jest.fn() }),
    useRoute: () => ({
        params: {}
    }),
}));



const spyAlert = jest.spyOn(Alert, 'alert');

const findElement = (tree, element) => {
    // console.log("Searched Tree: ", tree.children[0].children[1]);
    if (tree.props && tree.props.testID === element) {
    //   console.log("Searched Tree: ", tree);
      return tree;
    }
    if (tree.children && tree.children.length > 0) {
      let childs = tree.children;
      for (let i = 0; i < childs.length; i++) {
        let item = findElement(childs[i], element);
        if (typeof item !== "undefined") {
        //   console.log("Element: ", element);
          return item;
        }
      }
    }
  };

describe("Screen:: ParkingSlot ", () => {
    beforeEach(() => {
        store.dispatch(setParkingSlotList([]))
    })

    afterEach(() => {
        jest.clearAllMocks()
        cleanup()
    })

    test("Match snapshot ", () => {
        const navigation = { navigate: jest.fn() };
        store.dispatch(setParkingSlotList(mockDataSlots))
        const tree = render(
            <Provider store={store}>
                <ParkingSlot navigation={navigation} />
            </Provider>
        )
        expect(tree).toMatchSnapshot();
    })

    test("should render correctly ", async () => {
        const navigation = { navigate: jest.fn() };
        store.dispatch(setParkingSlotList(mockDataSlots))
        const { queryByTestId, getByTestId } = render(
            <Provider store={store}>
                <ParkingSlot navigation={navigation} />
            </Provider>
        )
        await waitFor(() => {
            return queryByTestId('post-row-0');
        });

        expect(getByTestId('post-row-0'));
    })

    test("Render empty list ", () => {
        const navigation = { navigate: jest.fn() };
        store.dispatch(setParkingSlotList([]))
        const tree = render(
            <Provider store={store}>
                <ParkingSlot navigation={navigation} />
            </Provider>
        )
        expect(tree).toMatchSnapshot();
    })

    test("Show empty message if no data ", async () => {
        const navigation = { navigate: jest.fn() };
        store.dispatch(setParkingSlotList([]))
        const { getByTestId } = render(
            <Provider store={store}>
                <ParkingSlot navigation={navigation} />
            </Provider>
        );
        await waitFor(() => {
            return getByTestId('empty-slots');
        });
        expect(getByTestId('empty-slots').children[0]).toBe(string.emptySlot);
    })

    test("Click on slot", () => {
        const mockedOnRemoveCar = jest.fn();
        const props: SlotContainerType = {
            index: 0,
            item: mockDataSlots[0],
            onRemoveCar: mockedOnRemoveCar
        }
        const { getByTestId } = render(<SlotContainer {...props} />)
        const rootRowId = getByTestId("post-row-0")
        fireEvent(rootRowId, 'onLongPress')
        expect(mockedOnRemoveCar).toHaveBeenCalled()
    })

    test("Remove Car: show alert: cancel ", async () => {
        store.dispatch(setParkingSlotList(mockDataSlots))
        const navigation = { navigate: jest.fn() };
        let tree = renderer.create(
            <Provider store={store}>
                <ParkingSlot navigation={navigation} />
            </Provider>).toJSON();
        expect(findElement(tree, "slot-flatlist")).toBeDefined();

        const FlatListRender = findElement(tree, "slot-flatlist").props.renderItem;
        console.log("props ",FlatListRender)

        const {getByTestId } = render(<FlatListRender index={0} item={mockDataSlots}/>)
        const toClick = getByTestId('post-row-0')
        fireEvent(toClick, 'onLongPress')
        // expect(a).toHaveBeenCalled()

        expect(Alert.alert).toHaveBeenCalled();
        // @ts-ignore
        spyAlert.mock.calls[0][2][0].onPress();     
    })

    test("Remove Car: show alert: ok ", async () => {
        store.dispatch(setParkingSlotList(mockDataSlots))
        const navigation = { navigate: jest.fn() };
        let tree = renderer.create(
            <Provider store={store}>
                <ParkingSlot navigation={navigation} />
            </Provider>).toJSON();
        expect(findElement(tree, "slot-flatlist")).toBeDefined();

        const FlatListRender = findElement(tree, "slot-flatlist").props.renderItem;
        console.log("props ",FlatListRender)

        const {getByTestId } = render(<FlatListRender index={0} item={mockDataSlots}/>)
        const toClick = getByTestId('post-row-0')
        fireEvent(toClick, 'onLongPress')
        // expect(a).toHaveBeenCalled()

        expect(Alert.alert).toHaveBeenCalled();
        // @ts-ignore
        spyAlert.mock.calls[0][2][1].onPress();     
    })

})

