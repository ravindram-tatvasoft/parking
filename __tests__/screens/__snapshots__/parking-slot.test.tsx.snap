// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen:: ParkingSlot  Match snapshot  1`] = `
<View
  style={
    Object {
      "flex": 1,
      "width": "100%",
    }
  }
>
  <View>
    <Text>
      Please long press on slot to remove car from parking slot
    </Text>
    <RCTScrollView
      data={
        Array [
          Object {
            "carNumber": "GJ 01 XX XXXX",
            "exitTime": "2022-01-25T17:59:42+05:30",
            "id": "id-1",
            "isEmpty": false,
            "slotNumber": 1,
            "time": 1,
          },
          Object {
            "carNumber": "GJ 02 XX XXXX",
            "exitTime": "2022-01-25T17:59:42+05:30",
            "id": "id-2",
            "isEmpty": true,
            "slotNumber": 2,
            "time": 2,
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      testID="slot-flatlist"
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <View
              accessible={true}
              focusable={false}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#d5d5d5",
                  "height": 355,
                  "justifyContent": "center",
                  "margin": 10,
                  "opacity": 1,
                  "width": 355,
                }
              }
              testID="post-row-0"
            >
              <Text
                style={
                  Array [
                    Object {
                      "textAlign": "center",
                    },
                    Object {
                      "color": "red",
                    },
                  ]
                }
              >
                Slot 1
 Not Available
              </Text>
              <View
                style={
                  Object {
                    "alignItems": "center",
                  }
                }
              >
                <Text>
                  GJ 01 XX XXXX
                </Text>
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 50,
                      "marginVertical": 10,
                      "width": 50,
                    }
                  }
                />
              </View>
            </View>
            <View
              accessible={true}
              focusable={false}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#d5d5d5",
                  "height": 355,
                  "justifyContent": "center",
                  "margin": 10,
                  "opacity": 1,
                  "width": 355,
                }
              }
              testID="post-row-1"
            >
              <Text
                style={
                  Array [
                    Object {
                      "textAlign": "center",
                    },
                    Object {
                      "color": "green",
                    },
                  ]
                }
              >
                Slot 2
 Available
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`Screen:: ParkingSlot  Render empty list  1`] = `
<View
  style={
    Object {
      "flex": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text
      testID="empty-slots"
    >
      Please create parking slot
    </Text>
  </View>
</View>
`;
